repos/system_upgrade/el8toel9/actors/checkdeprecatedrpmsignature/libraries/checkdeprecatedrpmsignature.py

```diff
@@ -0,0 +1,77 @@
+
+FMT_LIST_SEPARATOR = '\n    - '
+SUMMARY_FMT = (
+    'The SHA-1 hash algorithm is considered unsafe and it is no longer'
+    ' allowed to be used on RHEL 9 systems using the default crypto policy.'
+    ' This causes issues when using DNF/RPM to handle packages with RSA/SHA1'
+    ' signatures as the signature cannot be checked with the default crypto'
+    ' policy. Any such packages cannot be installed, removed,'
+    ' or replaced unless the signature check is disabled in dnf/rpm'
+    ' or SHA-1 is enabled using non-default crypto-policies.'
+    ' The list of problematic packages:{}'
+)
+
+# TODO(pstodulk): SHA1 sub policy does not exist on RHEL 8 and need to be created manually.
+# Regarding that, we now suggest only the LEGACY policy.
+REMEDY_HINT = (
+    'It is recommended to contact vendors of such packages to ask them for'
+    ' new builds with secure signatures and install the new packages before'
+    ' the upgrade. If this is not possible you may instead:'
+    ' a) remove such packages or b) update the crypto policy using the'
+    ' update-crypto-policies tool to allow SHA-1 usage on the upgraded'
+    ' system prior to the upgrade.'
+    ' In such a case the crypto policy must be set to "LEGACY".'
+)
+
+
+def _get_rpms_with_sha1_sig():
+    installed_rpms = next(api.consume(InstalledRPM)).items
+    return [pkg for pkg in installed_rpms if 'SHA1,' in pkg.pgpsig]
+
+
+def _is_sha1_allowed(current_policy):
+    """
+    Return True if we are sure the current policy allows SHA-1 on RHEL 9. False otherwise
+
+    Tee LEGACY policy and policies like DEFAULT:SHA1 enables SHA-1 on RHEL 9.
+    """
+    # TODO(pstodulk): this is just naive implementation
+    # NOTE: The SHA1 sub policy does not exist on RHEL 8. It has to be created
+    # manually by user in such a case.
+    # NOTE: for now, limit the check for :SHA1 to DEFAULT:SHA1 only as otherwise we will
+    # not be probably able to set correctly the policies inside the target
+    # userspace container
+    if current_policy == 'LEGACY' or "DEFAULT:SHA1" in current_policy:
+        return True
+    return False
+
+
+def process():
+    # TODO(pstodulk): add link to the official announce of the change in crypto policies
+    bad_rpms = _get_rpms_with_sha1_sig()
+    cpi = next(api.consume(CryptoPolicyInfo), None)
+    if bad_rpms:
+        bad_rpms_str = ''.join([
+            '{prefix}{pkgname} ({sig})'.format(prefix=FMT_LIST_SEPARATOR, pkgname=pkg.name, sig=pkg.pgpsig)
+            for pkg in bad_rpms
+        ])
+        report = [
+            reporting.Title('Detected RPMs with RSA/SHA1 signature'),
+            reporting.Summary(SUMMARY_FMT.format(bad_rpms_str)),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Tags([reporting.Tags.SECURITY, reporting.Tags.SANITY]),
+        ]
+        if not _is_sha1_allowed(cpi.current_policy):
+            # TODO:
+            cmds = [['update-crypto-policies', '--set', 'LEGACY']]
```

Comment:
What is the system that's affected with this and will it be reverted?

Reply:
@t184256  The host system (the system that is going to be upgraded). Regarding the changed remediation instruction, we will not print this CMD in the report. -- this command is not applied by us automatically. It's information for the user what command they could execute to deal with the issue. so as said, regarding the remediation instruction does not inform about legacy anymore, I will remove the cmd. DONE (with the next push, wait a minute..)